### ISSUE TITLE
Issue in instrumenting positional argument

### ISSUE THREAD

**Issue by nimamg** [sola-st/DynaPyt#55]
## Issue Description
The positional argument is not instrumented correctly. In the actual instrumented code, the result of the `join` function (which is a string) is given to the `_call_` function as the `call` argument (which is supposed to contain the callable object) instead of the `join` callable object itself, and the argument to the `join` function is not provided in the `pos_args` argument.
## Original Code
```python
".".join(str(x) for x in [1,2,3])
```
## Observed Instrumented Code
```python
 _rt._call_(
    _dynapyt_ast_, 8, 
    _rt._attr_(
        _dynapyt_ast_, 
        0, 
        ".", 
        "join"
    )(
        _rt._call_(
            _dynapyt_ast_, 
            2,
            str, 
            False, 
            [("", _rt._read_(_dynapyt_ast_, 1, lambda: x))], {}) for x in _rt._gen_(_dynapyt_ast_, 7, _rt._list_(_dynapyt_ast_, 6, [_rt._int_(_dynapyt_ast_, 3, 1),_rt._int_(_dynapyt_ast_, 4, 2),_rt._int_(_dynapyt_ast_, 5, 3)])
        )
    ), 
    True, 
    None, 
    None)
```
## Expected Instrumented Code
```python
 _rt._call_(
    _dynapyt_ast_, 8, 
    _rt._attr_(
        _dynapyt_ast_, 
        0, 
        ".", 
        "join"
    ), 
    True, 
    [("", _rt._call_(
        _dynapyt_ast_, 
        2,
        str, 
        False, 
        [("", _rt._read_(_dynapyt_ast_, 1, lambda: x))], {}) for x in _rt._gen_(_dynapyt_ast_, 7, _rt._list_(_dynapyt_ast_, 6, [_rt._int_(_dynapyt_ast_, 3, 1),_rt._int_(_dynapyt_ast_, 4, 2),_rt._int_(_dynapyt_ast_, 5, 3)])
    ))], 
    None)
```

**Comment by AryazE** [comment]
Thanks for raising this issue!
Feel free to reopen the issue if it is not fixed by 0d695fd.

**Comment by nimamg** [comment]
Thanks for the quick response! The issue with the generator expressions has been fixed, however I found another issue that I think is related, which is why I'm posting it here instead of opening a new issue.

## Issue Description
In some calls (I could reproduce it on builtin functions) the callable is called when being passed as the `call` argument.

## 1
### Original Code
```python
globals()
```
### Observed Instrumented Code
```python
_rt._call_(_dynapyt_ast_, 647, globals(), True, None, None)
```
### Expected Instrumented Code
```python
_rt._call_(_dynapyt_ast_, 647, globals, True, None, None)
```
## 2
### Original Code
```python
locals().keys()
```
### Observed Instrumented Code
```python
_rt._call_(_dynapyt_ast_, 646, _rt._attr_(_dynapyt_ast_, 645, _rt._call_(_dynapyt_ast_, 644, locals(), True, None, None), "keys"), False, [], {})
```
### Expected Instrumented Code
```python
_rt._call_(_dynapyt_ast_, 646, _rt._attr_(_dynapyt_ast_, 645, _rt._call_(_dynapyt_ast_, 644, locals, True, None, None), "keys"), False, [], {})
```
## 3
### Original Code
```python
vars(dict())
```
### Observed Instrumented Code
```python
_rt._call_(_dynapyt_ast_, 662, vars(_rt._call_(_dynapyt_ast_, 661, dict, False, [], {})), True, None, None)
```
### Expected Instrumented Code
```python
_rt._call_(_dynapyt_ast_, 662, vars, True, [("", _rt._call_(_dynapyt_ast_, 661, dict, False, [], {}))], None)
```

**Comment by AryazE** [comment]
Thanks! This is a special case that we have to call some functions at the original location in code.
Calling `locals` for example from the analysis code or the runtime engine results in different results from the original code.
For these function only the `post_call` hook is called, and there is an argument of the hook `post_only` that specifies such cases (if it is `True` it means that the function called was a special builtin).

**Comment by nimamg** [comment]
Calling the special functions in the code seems reasonable, however, I still have some issues with how it's done.
First, the `only_post` argument is not present in the hook, it's only present in the `_call_` function in the runtime engine, which does not provide any way for the analysis class to know if one of these special cases has happened.
Second, I believe it would still be beneficial if the callable object was present in the `call` argument, and the result was passed in a separate argument, with the `only_post` argument also added to the hook. Because in its present form, the `call` argument can either be a callable or not and it is up to the analysis class to use `try...except...` blocks to check for it. Although this would complicate matters a bit in the hook.

Whether my second suggestion is viable or not, I think that the `only_post` argument should be added to the hook anyways so that the analysis class can work around these special cases without relying on checking if the `call` argument is callable or not.

**Comment by AryazE** [comment]
Yes, you are right, I didn't notice it wasn't passed to the analyses.
That makes sense. I will update it soon.
It might take some time, since it will break some downstream code.

**Comment by AryazE** [comment]
Thank you for your patience!
I have looked into this, and it seems adding the `only_post` argument to the analysis hook would be redundant, as checking it would require the same amount of effort as checking `result is call`.
However, providing a consistent `Callable` argument would be more user-friendly. But I don't have a good idea on how to do this in a clean way. I am open to suggestions for this.
Nevertheless, I have added these to the documentation, and hope it prevents further mistakes.
